service/history/replication/executable_history_task.go

```diff
@@ -297,3 +302,145 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
+	}
+
+	if err := e.checkWorkflowKey(incomingHistoryTask.WorkflowKey); err != nil {
+		return nil, err
+	}
+
+	if err := e.checkVersionHistoryItem(incomingHistoryTask.versionHistoryItems); err != nil {
+		return nil, err
+	}
+
+	if err := e.checkBaseExecutionInfo(incomingHistoryTask.baseExecutionInfo); err != nil {
+		return nil, err
+	}
+
+	events, newRunEvents, err := incomingHistoryTask.getDeserializedEvents()
+	if err != nil {
+		return nil, err
+	}
+
+	if err = e.checkEvents(events, newRunEvents); err != nil {
+		return nil, err
+	}
+
+	return incomingHistoryTask, nil
+}
+
+// checkVersionHistoryItem will check if incoming tasks Version history is on the same branch as the current one
+func (e *ExecutableHistoryTask) checkVersionHistoryItem(incomingHistoryItems []*historyspb.VersionHistoryItem) error {
+	if len(e.versionHistoryItems) == 0 || len(incomingHistoryItems) == 0 {
+		return serviceerror.NewInvalidArgument("version history empty")
+	}
+
+	firstNotEqualIndex := 0
+	for firstNotEqualIndex < len(e.versionHistoryItems) && firstNotEqualIndex < len(incomingHistoryItems) &&
+		e.versionHistoryItems[firstNotEqualIndex].Equal(incomingHistoryItems[firstNotEqualIndex]) {
+		firstNotEqualIndex++
+	}
+	currentLastIndex := len(e.versionHistoryItems) - 1
+	currentLastItem := e.versionHistoryItems[currentLastIndex]
+	if firstNotEqualIndex == len(e.versionHistoryItems) || // current version history is part of incoming version history
+		(firstNotEqualIndex == currentLastIndex &&
+			currentLastItem.Version == incomingHistoryItems[firstNotEqualIndex].Version &&
+			currentLastItem.EventId <= incomingHistoryItems[firstNotEqualIndex].EventId) {
+		return nil
+	}
+
+	return serviceerror.NewInvalidArgument("version history does not match")
+}
+
+func (e *ExecutableHistoryTask) checkWorkflowKey(incomingWorkflowKey definition.WorkflowKey) error {
+	if !(e.WorkflowKey.GetWorkflowID() == incomingWorkflowKey.GetWorkflowID() &&
+		e.WorkflowKey.GetNamespaceID() == incomingWorkflowKey.GetNamespaceID() &&
+		e.WorkflowKey.GetRunID() == incomingWorkflowKey.GetRunID()) {
+		return serviceerror.NewInvalidArgument("workflow key does not match")
+	}
+	return nil
+}
+
+func (e *ExecutableHistoryTask) checkBaseExecutionInfo(incomingTaskExecutionInfo *workflowpb.BaseExecutionInfo) error {
+	if e.baseExecutionInfo != nil && incomingTaskExecutionInfo != nil && !e.baseExecutionInfo.Equal(incomingTaskExecutionInfo) {
+		return serviceerror.NewInvalidArgument("base execution info not match")
+	}
+
+	return nil
+}
+
+func (e *ExecutableHistoryTask) checkEvents(incomingEvents []*historypb.HistoryEvent, incomingNewRunEvents []*historypb.HistoryEvent) error {
+	if len(incomingEvents) == 0 {
+		return serviceerror.NewInvalidArgument("incoming task is empty")
+	}
+	currentEvents, currentNewRunEvents, err := e.getDeserializedEvents()
+	if err != nil {
+		return err
+	}
+
+	if currentNewRunEvents != nil && incomingNewRunEvents != nil {
```

Comment:
If current new run !=nil, are we expected the task is not batchable?

Reply:
Yeah, we should not expect the task is batchable because `currentNewRun != nil` indicating the current task is the last task. I will change the logic safer.